Add support for param mutation under inference mode

tugsbayasgalan opened the pull request:
Summary:
In HF model rwkv, we have parameter mutation under inference mode which should be safe. This PR does multiple things to make sure it works:
1. We execute global autograd mutation while tracing so that we can actually trace through parameter inplace mutation 
2. Add support for parameter mutation under inference mode in AOTAutograd 
3. Add support for parameter mutation under inference mode in export.

Test Plan:
test

Rollback Plan:

Differential Revision: D79460136




cc @ezyang @SherlockNoMad @EikanWang @jgong5 @wenzhe-nrv

tugsbayasgalan commented on torch/_functorch/_aot_autograd/input_output_analysis.py:
```diff
@@ -460,6 +460,7 @@ def create_graph_signature(
         named_buffers=buffer_names,
         num_user_inputs=num_user_args,
         num_user_outputs=num_user_fw_outs,
+        trace_joint=trace_joint,
```
cc: @ezyang does this look ok in terms of your refactoring work?

ezyang replied:
The change looks reversible enough, but I can't really say I understand why you need yet another flag here. Additionally in your tests you are not even tracing out the joint!

ezyang replied:
One thing I'll say is that for a generic "this works when torch.compile works" export frontend, these cases also need to be accounted for.

tugsbayasgalan commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -1430,9 +1430,11 @@ def __torch_function__(
                 node.meta["val"] = None
+            # For autocast, the python APIs run so we don't have to run them again 
+            # here. 
+            if func is torch._C._set_grad_enabled:
```
cc: @zou3519, @bdhirsh This is behavioral change from before.

zou3519 replied:
So, at the very least, pre-dispatch tracing should not change the state after it runs. If there is only one single torch.set_grad_enabled(False) call in the graph, after one does the pre-dispatch trace, it should not be off.

tugsbayasgalan replied:
Yep i wrapped it in ctx manager under export.

zou3519 reviewed (commented):
your tests don't look happy

ydwu4 commented on torch/export/graph_signature.py:
```diff
@@ -121,6 +121,7 @@ class OutputKind(Enum):
     USER_OUTPUT = auto()
     LOSS_OUTPUT = auto()
     BUFFER_MUTATION = auto()
+    PARAMETER_MUTATION = auto()
```
Does this gets properly serialized and de-serialized? Curious why this doesn't trigger the error [here](https://github.com/pytorch/pytorch/blob/main/torch/_export/serde/serialize.py#L1279).  We need to add the serialize and deserialize logic as well.

BC-wise, this could be made BC-preserving i.e. a minor version upgrade. The idea is that add a new  optional field parameter_mutation in scehma.OutputSpec.  Deserializer should be automatically assign it None when de-serializing old artifact.

tugsbayasgalan commented on test/export/test_export.py:
```diff
@@ -325,6 +325,36 @@ def forward(self, *args):
+        self.assertTrue("parameter" in val.values())
+
+        with self.assertRaisesRegex(RuntimeError, "leaf"):
+            ep.module()(torch.rand(4, 4))
```
Yeah we can do that

ydwu4 commented on torch/export/_trace.py:
```diff
@@ -1674,7 +1674,24 @@ def override_getattribute_for_subclasses(args):
+            with (
+                ctx,
+                override_getattribute_for_subclasses(flat_args),
+                _maybe_restore_grad_state(),
```
Is it accurate that parameter mutation is supported in export (for all the flags/mode)? Might worth document all the semantic of mutations, e.g. input, buffer, parameters somewhere. Right now I don't have a clear mental model for them.

tugsbayasgalan committed:
Add support for param mutation under inference mode (#159661)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/159661

In HF model rwkv, we have parameter mutation under inference mode which should be safe. This PR does multiple things to make sure it works:
1. We execute global autograd mutation while tracing so that we can actually trace through parameter inplace mutation
2. Add support for parameter mutation under inference mode in AOTAutograd
3. Add support for parameter mutation under inference mode in export.

Test Plan:
test

Rollback Plan:

Reviewed By: ydwu4

Differential Revision: D79460136
